metautils/lib/gridd_client.c

```diff
@@ -825,22 +824,107 @@ gridd_client_fail(struct gridd_client_s *client, GError *why)
 
 /* ------------------------------------------------------------------------- */
 
-void
-gridd_client_learn_peers_down(const char * const * peers)
+static gchar*
+_srv_addr(struct service_info_s *si)
+{
+	EXTRA_ASSERT(si != NULL);
+	gchar addr[STRLEN_ADDRINFO];
+	grid_addrinfo_to_string(&si->addr, addr, sizeof(addr));
+	return g_strdup(addr);
+}
+
+static gboolean
+_srv_is_down(struct service_info_s *si)
 {
-	GTree *new_down = NULL, *old_down = NULL;
+	EXTRA_ASSERT(si != NULL);
 
-	new_down = g_tree_new_full(metautils_strcmp3, NULL, g_free, NULL);
-	for (const char * const * ppeer = peers; peers && *ppeer ;++ppeer)
-		g_tree_replace(new_down, g_strdup(*ppeer), (void*)0xDEADBEAF);
+	if (!si->tags)
+		return FALSE;
+
+	struct service_tag_s *tag =
+		service_info_get_tag(si->tags, NAME_TAGNAME_UP);
+	if (!tag)
+		return FALSE;
```

Comment:
Please add a comment something like
Legacy: consider service up if tag is missing